src/shogun/machine/gp/GaussianARDFITCKernel.cpp

```diff
@@ -0,0 +1,146 @@
+{
+	init();
+}
+void CGaussianARDFITCKernel::init()
+{
+}
+
+CGaussianARDFITCKernel::~CGaussianARDFITCKernel()
+{
+}
+
+#ifdef HAVE_CXX11
+#include <shogun/mathematics/linalg/linalg.h>
+CGaussianARDFITCKernel::CGaussianARDFITCKernel(int32_t size, float64_t width)
+		: CGaussianARDKernel(size,width)
+{
+	init();
+}
+
+CGaussianARDFITCKernel::CGaussianARDFITCKernel(CDotFeatures* l,
+		CDotFeatures* r, int32_t size, float64_t width)
+		: CGaussianARDKernel(l, r, size, width)
+{
+	init();
+}
+
+CGaussianARDFITCKernel* CGaussianARDFITCKernel::obtain_from_generic(CKernel* kernel)
+{
+	if (kernel->get_kernel_type()!=K_GAUSSIANARDFITC)
+	{
+		SG_SERROR("Provided kernel is not of type CGaussianARDFITCKernel!\n");
+	}
+
+	/* since an additional reference is returned */
+	SG_REF(kernel);
+	return (CGaussianARDFITCKernel*)kernel;
+}
+
+#include <typeinfo>
+CSGObject *CGaussianARDFITCKernel::shallow_copy() const
+{
+	// TODO: remove this after all the classes get shallow_copy properly implemented
+	// this assert is to avoid any subclass of CGaussianARDFITCKernel accidentally called
+	// with the implement here
+	ASSERT(typeid(*this) == typeid(CGaussianARDFITCKernel))
+	CGaussianARDFITCKernel *ker = new CGaussianARDFITCKernel(cache_size, m_width);
+	if (lhs)
+	{
+		ker->CGaussianARDKernel::init(lhs, rhs);
+	}
+	return ker;
+}
+
+SGMatrix<float64_t> CGaussianARDFITCKernel::get_parameter_gradient(
+		const TParameter* param, index_t index)
+{
+	if (!strcmp(param->m_name, "latent_features"))
+	{
+		REQUIRE(lhs && rhs, "Features not set!\n")
+		REQUIRE(index>=0 && index<num_lhs,"")
+		int32_t idx_l=index;
+		SGVector<float64_t> left_vec=((CDotFeatures *)lhs)->get_computed_dot_feature_vector(idx_l);
+		SGMatrix<float64_t> res(left_vec.vlen, num_rhs);
+
+		for (int32_t idx_r=0; idx_r<num_rhs; idx_r++)
+		{
+			SGVector<float64_t> right_vec=((CDotFeatures *)rhs)->get_computed_dot_feature_vector(idx_r);
+			SGMatrix<float64_t> res_transpose(res.get_column_vector(idx_r),1,left_vec.vlen,false);
+			linalg::add(left_vec, right_vec, right_vec, 1.0, -1.0);
+			float64_t scalar_weight=1.0;
+			//column vector
+			SGMatrix<float64_t> right=compute_right_product(right_vec, scalar_weight);
+			//row vector
+			SGMatrix<float64_t> right_transpose(right.matrix,1,right.num_rows,false);
+			if (m_ARD_type==KT_SCALAR)
+			{
+				scalar_weight*=m_weights[0];
+				SGMatrix<float64_t> weights(1,right.num_rows);
+				weights.set_const(scalar_weight);
+				linalg::matrix_product(right_transpose, weights, res_transpose);
+			}
+			else
+			{
+				if(m_ARD_type==KT_DIAG)
+				{
+					SGMatrix<float64_t> weights(m_weights.matrix,1,m_weights.num_rows,false);
+					linalg::elementwise_product(right_transpose, weights, res_transpose);
+				}
+				else if(m_ARD_type==KT_FULL)
+				{
+					linalg::matrix_product(right_transpose, m_weights, res_transpose);
+				}
+				else
+				{
+					SG_ERROR("Unsupported ARD kernel\n");
```

Comment:
The error message should be more clear. "Unsupported ARD type, see <name_of_enum>\"